### ISSUE TITLE
False positive on match_same_arms

### ISSUE THREAD

**Issue by clarcharr** [rust-lang-nursery/rust-clippy#1996]
Basically doesn't pay attention to `if`s in arms before the wildcard.

Example code:

```rust
// validate months and days
match (month, day) {
    // remove days that never exist
    (2, 30) | (2, 31) |
    (4, 31) |
    (6, 31) |
    (9, 31) |
    (11, 31) => (),

    // remove leap day on non-leap-years
    (2, 29) if year % 4 != 0 || year % 100 == 0 || year % 400 != 0 => (),

    // ensure valid month/day
    (1...12, 1...31) => /* parse y/m/d */,

    _ => (),
}
```
